lib/scholar/decomposition/incremental_pca.ex

```diff
@@ -0,0 +1,311 @@
+           containers: [
+             :components,
+             :singular_values,
+             :num_samples_seen,
+             :mean,
+             :variance,
+             :explained_variance,
+             :explained_variance_ratio
+           ]}
+  defstruct [
+    :components,
+    :singular_values,
+    :num_samples_seen,
+    :mean,
+    :variance,
+    :explained_variance,
+    :explained_variance_ratio,
+    :whiten?
+  ]
+
+  opts = [
+    num_components: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of principal components."
+    ],
+    whiten?: [
+      type: :boolean,
+      default: false,
+      doc: """
+      When true the `components` are divided by `num_samples` times `components` to ensure uncorrelated outputs with unit component-wise variances.
+
+      Whitening will remove some information from the transformed signal (the relative variance scales of the components)
+      but can sometimes improve the predictive accuracy of the downstream estimators by making data respect some hard-wired assumptions.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits an Incremental PCA model on a stream of batches.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return values
+
+  The function returns a struct with the following parameters:
+
+    * `:num_components` - The number of principal components.
+
+    * `:components` - Principal axes in feature space, representing the directions of maximum variance in the data.
+      Equivalently, the right singular vectors of the centered input data, parallel to its eigenvectors.
+      The components are sorted by decreasing `:explained_variance`.
+
+    * `:singular_values` - The singular values corresponding to each of the selected components.
+      The singular values are equal to the 2-norms of the `:num_components` variables in the lower-dimensional space.
+
+    * `:num_samples_seen` - The number of data samples processed.
+
+    * `:mean` - Per-feature empirical mean.
+
+    * `:variance` - Per-feature empirical variance.
+
+    * `:explained_variance` - Variance explained by each of the selected components.
+
+    * `:explained_variance_ratio` - Percentage of variance explained by each of the selected components.
+
+    * `:whiten?` - Whether to apply whitening.
+
+  ## Examples
+
+      iex> {x, _} = Scidata.Iris.download()
+      iex> batches = x |> Nx.tensor() |> Nx.to_batched(10)
+      iex> ipca = Scholar.Decomposition.IncrementalPCA.fit(batches, num_components: 2)
+      iex> ipca.components
+      Nx.tensor(
+        [
+          [-0.33354005217552185, 0.1048964187502861, -0.8618107080105579, -0.3674643635749817],
+          [-0.5862125754356384, -0.7916879057884216, 0.15874788165092468, -0.06621300429105759]
+        ]
+      )
+      iex> ipca.singular_values
+      Nx.tensor([77.05782028025969, 10.137848854064941])
+  """
+  def fit(batches = %Stream{}, opts) do
```

Comment:
We should not match on a stream because `%Stream{}` is only one of the possible streams. Instead we should just treat it as an enumerable:

```suggestion
  def fit(batches, opts) do
```

Reply:
I wonder though if we should instead merge this into PCA and add a new function called `fit_stream`.

Reply:
Yes, I was about to recommend that! Do we need to call it `fit_stream` or `fit_incremental` or we can just add another clause? Two clauses were working over here; one was defined with `deftransform`, the other with `def`.

Reply:
I would do a separate function. Either fit_stream or fit_incremental.